File: src/main/kotlin/SerialInterface.java
import com.fazecast.jSerialComm.SerialPort;
import com.fazecast.jSerialComm.SerialPortEvent;
import com.fazecast.jSerialComm.SerialPortMessageListener;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

public class SerialInterface implements SerialPortMessageListener{
    private static SerialPort port           = null;   // The serial port
    public static List<serial_callback> callbacks = new ArrayList<>();
    public static String delimiter = "\r\n";

    private static boolean is_opened = false;

    public boolean is_opened() {
        return is_opened;
    }

    /**
     * @return Get current port's descriptive port name.
     */
    public String port_name() {
        try{
            return port.getDescriptivePortName();
        } catch (Exception ignored) {};
        return null;
    }

    SerialInterface() {
        System.setProperty("fazecast.jSerialComm.appid", "QuokeCola.Serial-Terminal");
    }

    /**
     * Get the available serial ports through jSerialComm
     */
    SerialPort[] get_ports() {
        return SerialPort.getCommPorts();
    }

    void set_port_baudrate(int baudrate) {
        port.setBaudRate(baudrate);
    }

    int get_port_baudrate() {
       return port.getBaudRate();
    }

    /**
     * Set the serial port for this serial interface and open it.
     * @param port_ Available port that obtained from <SerialInterface.get_ports()> method
     */
    void open_port(SerialPort port_) {
        try{ // if port is not null.
            stop_receive();
        } catch (Exception ignored) { // Initially, port is not set yet so there will be a null error
        }
        // Set port.
        port = port_;
        port.allowElevatedPermissionsRequest();
        port.openPort();
        port.setComPortTimeouts(SerialPort.TIMEOUT_NONBLOCKING, 1000, 0);
        port.addDataListener(this);
        is_opened = true;
    }

    /**
     * Stop receiving data.
     */
    void stop_receive() {
        if (port.isOpen()) {
            port.closePort();
            port.removeDataListener();
        }
        is_opened = false;
    }

    static void send(String string, String delimiter) throws IOException {
        byte[] out = (string + delimiter).getBytes();
        port.getOutputStream().write(out);
    }

    @Override
    public byte[] getMessageDelimiter() {
        return delimiter.getBytes();
    }

    @Override
    public boolean delimiterIndicatesEndOfMessage() {
        return false;
    }

    @Override
    public int getListeningEvents() {
        return SerialPort.LISTENING_EVENT_DATA_RECEIVED;
    }

    @Override
    public void serialEvent(SerialPortEvent event) {
        byte[] rx_byte = event.getReceivedData();
        String rx_str  = new String(rx_byte, StandardCharsets.ISO_8859_1);
        rx_str = rx_str.strip();
        if (!rx_str.equals("")) {
            for (serial_callback callback:
                    callbacks) {
                try {
                    callback.process_callback(rx_str.strip());
                } catch (Exception ignored) {}
            }
        }
    }
}

File: src/main/kotlin/SerialInterface.kt
@file:Suppress("PrivatePropertyName", "PropertyName")

import com.fazecast.jSerialComm.SerialPort


class SerialInterface: Thread() {
    private val BUFFER_SIZE: Int = 10000                  // MAXIMUM LINES THE BUFFER COULD STORE
    private var port : SerialPort? = null            // The serial port
    private var input_stream  = port?.inputStream    // IO input  stream
    private var output_stream = port?.outputStream   // IO output stream

    private var buffer_string = ""
    var buffer_line = mutableListOf<String>("WELCOME TO SERIAL TERMINAL", "AUTHOR: CHEN QIAN")

    /**
     * Initialize the SerialInterface
     */
    init {
        System.setProperty("fazecast.jSerialComm.appid", "ChibiOS-Terminal")
    }

    /**
     * Get the available serial ports through jSerialComm
     */
    fun get_ports(): Array<SerialPort> {
        return SerialPort.getCommPorts();
    }

    /**
     * Set the serial port for this serial interface
     * @param port_ Available port that obtained from <SerialInterface.get_ports()> method
     */
    fun set_port(port_: SerialPort) {
        port = port_
    }

    /**
     * Start the data receive thread.
     */
    override fun start() {
        port?.openPort()
        port?.setComPortTimeouts(SerialPort.TIMEOUT_READ_SEMI_BLOCKING, 0, 0)
        input_stream = port?.inputStream
        super.start()
    }

    /**
     * Data receive thread.
     */
    override fun run() {
        try {
            // Added buffer string at each iteration
            buffer_string += input_stream?.read()?.toChar()
            // If buffer contains /r or /n, stores the lines into buffer lines and remove the first line.
            if (buffer_string.contains("/r")||buffer_string.contains("/n")){
                buffer_line.add(buffer_string.split("/n")[0].removeSurrounding("/r"))
                buffer_string = buffer_string.split("/n")[1].removeSurrounding("/r")
                if (buffer_line.size > BUFFER_SIZE) {
                    buffer_line.drop(1);
                }
            }

        } catch (e: Exception) {
            // If error occurs
            e.printStackTrace()
            port?.closePort()

        }
    }

}